### ISSUE TITLE
Often, when I open and save a scene that has nodes assigned to exported node properties, the reference is replaced with an instance in the scene source

### ISSUE THREAD

**Issue by sandord** [godotengine/godot#86632]
### Tested versions

- Reproducable in v4.2.1.stable.mono.official [b09f793f5]
- I've seen this happen in versions of Godot 4 as well

### System information

Windows 11 - v4.2.1.stable.mono.official [b09f793f5] - Mobile renderer

### Issue description

Often, when I open and save a scene (without changing anything manually) that has nodes assigned to exported node properties, references are replaced with an instance in the scene source. The following diff screenshot should illustrate what I mean:

![image](https://github.com/godotengine/godot/assets/260737/44a0eeb6-5734-4455-a96e-a0a418ac8953)

This is a short version of the `Teleport` script, which is behind the affected nodes:

```csharp
[Tool]
public partial class Teleport : PanelContainer
{
    [Export]
    public Teleport? Destination { get; set; }
}
```

The described problem doesn't happen all the time but more often than not.

Also, I've seen similar things happen in my project that result in node references becoming severed. Many times, my game crashes for me to find out that a node has lost is reference to the Player node again and I have to hit Assign in the editor to re-attach it.

### Steps to reproduce

Open and immediately save a scene with nodes in it that have scripts with exported properties that point to other nodes in the scene. There's probably much more to it that I can't yet grasp.

### Minimal reproduction project (MRP)

I've set up a minimal reproduction project I haven't been able to put in the right things to trigger the bug yet. My original project is way too large to post here.

**Comment by AlexOtsuka** [comment]
Hello sandord,

I haven't been able to reproduce this bug after about an hour of messing around with it and trying different configurations using PackedScenes, different levels of encapsulation, etc. I myself use [Export] extensively in my C# projects, and haven't encountered this bug.

I'm aware of an issue I had in the past, where naming an ExportGroup the same as one of its properties would break the reference when building the scene (not a bug, but a consequence of how ExportGroups are implemented), e.g.: 
```
[ExportGroup("Health")]
[Export] public Health Health { get; private set; }
```

A few things to note:

- You don't need the ? here, as reference types are already nullable.
- Exports can be made private, or public with a private setter for better encapsulation. if they don't require being accessed or modified by other scripts. Maybe another script somewhere is affecting the value here?
- I'm wondering about your use of the [Tool] attribute here, and whether code you're executing in the editor could be contributing to this bug. Node that [Tool] isn't needed for [Export] to work.

I'd love to look more into it if you're able to find a way to come up with a working MRP or if you come up with an idea as to what could be causing this.

**Comment by sandord** [comment]
> I haven't been able to reproduce this bug after about an hour of messing around with it and trying different configurations using PackedScenes, different levels of encapsulation, etc. I myself use [Export] extensively in my C# projects, and haven't encountered this bug.

First of all, thanks for looking into this!

> I'm aware of an issue I had in the past, where naming an ExportGroup the same as one of its properties would break the reference when building the scene (not a bug, but a consequence of how ExportGroups are implemented), e.g.:
> 
> ```
> [ExportGroup("Health")]
> [Export] public Health Health { get; private set; }
> ```
 
> A few things to note:
> 
> * You don't need the ? here, as reference types are already nullable.

It's because I have `<nullable>enabled</nullable>` in my `.csproject` file. If I omit the `?`, the compiler will start issuing warnings and I have grown accustomed to having explicit nullability enabled as it helps finding bugs more easily.

> * Exports can be made private, or public with a private setter for better encapsulation. if they don't require being accessed or modified by other scripts. Maybe another script somewhere is affecting the value here?

Hmm, that's something I have forgotten about. I'm sure there are no scripts writing to the properties but I'm going to make them private anyway. Thanks for pointing it out.

> * I'm wondering about your use of the [Tool] attribute here, and whether code you're executing in the editor could be contributing to this bug. Node that [Tool] isn't needed for [Export] to work.

Yeah, I should have removed that from the example too. It's from my project where I've overridden `_GetConfigurationWarnings` to warn about a missing destination in the Editor.

> I'd love to look more into it if you're able to come up with a working MRP or if you have an idea as to what could be causing this.

I just made the `Destination` setter on the teleport `private` (no script was writing to it anyway) so let's see what difference it makes. I've tried saving the scene a f ew times and the problem hasn't shown up yet. Fingers crossed.

**Comment by sandord** [comment]
I haven't been able to work on my project during the holidays but I just tried opening, editing and saving the most notorious scenes and I haven't seen any unwanted changes in my diffs so far. I've been in a similar situation before so I'm not ready to draw any conclusions yet.

**Comment by AThousandShips** [comment]
Is the nose instance in the tree? If it's not a node attached to the tree it will be saved directly like this

**Comment by sandord** [comment]
> Is the nose instance in the tree? If it's not a node attached to the tree it will be saved directly like this

Yes, this involves several nodes that reference each other (teleports). They are all in the tree (added inside the editor and in no way modified by any scripts that I wrote so they are definitely not detached.

**Comment by dsnopek** [comment]
I have also seen this happen, although, pretty rarely, and I have no idea how to reproduce it on purpose. In my case, it was with a GDScript (not C#) project.

**Comment by Cronos87** [comment]
I'm running into this issue as well with exported nodes. The scenes that add a scene that contrains exported nodes will have all the exported properties duplicate in the scene file.

For example:

I have a scene MyLevel that includes Menu. Menu has 2 exported properties that contains nodes. Once I save a scene, both exported properties will be included in MyLevel which resulsts of duplicating data.

It is important to notice that MyLevel must be opened in the editor, if not, there is no issue.

I have no access to a computer to provide a testable project, as soon as I can I will try to provide one.

By the way, I'm using GDScript.

**Comment by AlexOtsuka** [comment]
I can confirm that the observation @Cronos87 has posted about happens on 4.3 and not on 4.2.1.

In 4.2.1: The parent scene file will not contain the child scene's exported property's value unless it's been changed from its default value.

In 4.3: The parent scene file will always contain the child scene's exported property's value even if it has not been changed.

I have no idea whether this is a bug, or a desirable feature that's being worked on.
However, this seems like a completely different problem from what @sandord was dealing with, which is scenes losing references to exported nodes, causing their values to be set to null. As far as I can tell, this doesn't break anything and only increases the size of scene files.

**Comment by sandord** [comment]
Today, it happened again.

**Comment by AThousandShips** [comment]
@sandord Please don't bump without contributing significant new information.

**Comment by Cronos87** [comment]
Sorry, English isn't my mother tongue, I may didn't understand well the main issue.

Anyway, regarding the main issue, I often ran into it, and I noticed it happens when I edit a script and have an error, like a parsing error. In this case, all exported variables will deseappears from the inspector, and I suspect the issue to appear at this precise moment.

I am not sure at 100% that the way to trigger the issue, this is based on my experience, and I can't confirm it at the moment. As soon as I can I will test my theorie.

**Comment by sandord** [comment]
Another example of the issue happening, with a lot of 'Max recursion reached' errors in the editor being displayed.

I didn't change anything to the teleports myself, I just moved an unrelated node in the scene.

![image](https://github.com/godotengine/godot/assets/260737/f42575fb-0f6e-4dc7-af85-237a1dfc8468)

![image](https://github.com/godotengine/godot/assets/260737/1fd5d326-fae9-4999-99d3-56b5e5b59411)

```csharp
[Tool]
public partial class Teleport : PanelContainer
{
    [Export]
    public Teleport? Destination { get; private set; }

    public override void _Ready()
    {
        if (!Engine.IsEditorHint())
        {
            Hide();
        }
    }
```

**Comment by sandord** [comment]
I just added the following editor hint and I'm under the impression that the scene is not corrupted anymore when saving it:

```csharp
[Export(PropertyHint.NodeType, hintString: "Teleport")]
public Teleport? Destination { get; private set; }
```
